returnn/training.py

```diff
@@ -573,3 +573,179 @@ def hash(cls, kwargs):
+        self.learning_rates = learning_rates
+        self.index = index
+        self.out_epoch = self.output_var("epoch")
+        self.key = key
+
+    def run(self):
+        # this has to be defined in order for "eval" to work
+        def EpochData(learningRate, error):
+            return {"learning_rate": learningRate, "error": error}
+
+        with open(self.learning_rates.get_path(), "rt") as f:
+            text = f.read()
+
+        data = eval(
+            text, {"nan": float("nan"), "inf": float("inf"), "EpochData": EpochData}
+        )
+
+        epochs = list(sorted(data.keys()))
+
+        # some epochs might not have all keys, we require that the last entry has the key, other epochs which
+        # do not have it might be ignored
+        available_keys = data[epochs[-1]]["error"]
+        if self.key not in available_keys:
+            raise KeyError(
+                f"{self.key} is not available in the provided learning_rates file f{self.learning_rates.get_path()}"
+            )
+
+        scores = [
+            (epoch, data[epoch]["error"][self.key])
+            for epoch in epochs
+            if self.key in data[epoch]["error"]
+        ]
+        sorted_scores = list(sorted(scores, key=lambda x: x[1]))
+
+        self.out_epoch.set(sorted_scores[self.index][0])
+
+    def tasks(self):
+        yield Task("run", mini_task=True)
+
+
+class GetBestTFCheckpointJob(GetBestEpochJob):
+    """
+    Returns the best checkpoint given a training model dir and a learning-rates file
+    The best checkpoint will be HARD-linked if possible, so that no space is wasted but also the model not
+    deleted in case that the training folder is removed.
+    """
+
+    def __init__(
+        self, model_dir: tk.Path, learning_rates: tk.Path, key: str, index: int = 0
+    ):
+        """
+
+        :param Path model_dir: model_dir output from a RETURNNTrainingJob
+        :param Path learning_rates: learning_rates output from a RETURNNTrainingJob
+        :param int index: index of the sorted list to access, 0 for the lowest, -1 for the highest score
+        :param str key: a key from the learning rate file that is used to sort the models
+            e.g. "dev_score_output/output_prob"
+        """
+        super().__init__(model_dir, learning_rates, key, index)
+        self._out_model_dir = self.output_path("model", directory=True)
+
+        # Note: checkpoint.index (without epoch number) is only allowed be cause we are symlinking
+        # to a file that contains an epoch number, as RETURNN will resolve symlinks automatically and
+        # then check for the epoch number
+        self.out_checkpoint = Checkpoint(self.output_path("model/checkpoint.index"))
+
+    def tasks(self):
+        yield Task("run", mini_task=True)
+
+    def run(self):
+        super().run()
+
+        try:
+            os.link(
+                os.path.join(
+                    self.model_dir.get_path(), "epoch.%.3d.index" % self.out_epoch.get()
```

Comment:
do you think hardcoding `"epoch.%.3d" % epoch` is general enough? 
What if we have a pre-train epoch or someone renamed the default name in their custom training job?

Reply:
Ah yes I forgot about pretraining. We will probably not have that anymore in newer configs (because via `get_network` mechanism, you can also do pretraining, but models checkpoints are always called the same). But it would be nice to support them as well here. I would maybe add an argument `last_pretrain_epoch=0` or so.

Reply:
"Modern" pre-training logic does not alter the name any more... And I would say the first person who actually has a custom training job can add a new parameter if needed...

Reply:
Ah we answered simultaneously :) I think we can add such a parameter if actually needed. I think those jobs will be used for rather new setups. Also I never had the case (yet) in the models using the old pre-training logic went that far that the checkpoint was actually considered for "best selection".

Reply:
ok, then I think we can keep it this way